src/main.rs

```diff
@@ -1,3 +1,207 @@
-fn main() {
+use chrono::NaiveDateTime;
+use clap::Parser;
+use serde::{Deserialize, Serialize};
 
+use rayon::{prelude::*, current_num_threads};
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::path::PathBuf;
+use std::{fs, io};
+
+fn main() -> io::Result<()> {
+    let cli = Cli::parse();
+
+    if let Some(jobs) = cli.jobs {
+        rayon::ThreadPoolBuilder::new().num_threads(jobs).build_global().unwrap();
+    }
+
+    let content = fs::read_to_string(cli.file)?;
+
+    let chat: Chat = serde_json::from_str(&content).unwrap();
+    let stat: ChatStatistics = ChatStatistics::gather(&chat);
+
+    let file = fs::File::create(cli.output)?;
+    serde_json::to_writer_pretty(file, &stat)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Parser)]
+struct Cli {
+    #[arg(long, short)]
+    file: PathBuf,
+    #[arg(long, short, default_value = "out.json")]
+    output: PathBuf,
+    #[arg(long, short)]
+    jobs: Option<usize>
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Person<'a>(Cow<'a, str>);
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Token<'a>(Cow<'a, str>);
+
+impl<'a> From<&'a str> for Token<'a> {
+    fn from(value: &'a str) -> Self {
+        Token(value.into())
+    }
+}
+impl<'a> From<String> for Token<'a> {
+    fn from(value: String) -> Self {
+        Token(value.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Chat<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    chat_type: ChatType,
+    id: u128,
+    messages: Vec<Message<'a>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum ChatType {
+    PublicChannel,
+    PrivateChannel,
+    PublicSupergroup,
+    PrivateSupergroup,
+    PersonalChat,
+    ChatForbidden,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Message<'a> {
+    id: u64,
+    #[serde(rename = "type")]
+    message_type: MessageType,
+    date: NaiveDateTime,
+    from: Option<Person<'a>>,
+    text_entities: Vec<TextEntity>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+enum MessageType {
+    Service,
+    Message,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TextEntity {
+    #[serde(rename = "type")]
+    text_type: TextEntityType,
+    text: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+enum TextEntityType {
+    Pre,
+    Bold,
+    Link,
+    Code,
+    Email,
+    Plain,
+    Phone,
+    Italic,
+    Cashtag,
+    Spoiler,
+    Mention,
+    Hashtag,
+    TextLink,
+    Underline,
+    BotCommand,
+    CustomEmoji,
+    MentionName,
+    Strikethrough,
+}
+
+impl TextEntityType {
+    pub fn is_meta(&self) -> bool {
+        use TextEntityType::*;
```

Comment:
I'm doing it inside a function, it's okay on a local level

Reply:
It's not about locality. It's about extracting everything. You might extend your `TextEntityType` in the future, so that unnecessary/useless struct/enums/etc will be imported without noticing it.

Reply:
> It's not about locality. It's about extracting everything. You might extend your TextEntityType in the future, therefore unnecessary/useless structs/enums/etc will come out of the blue.

And what are side effects? On global level it may be namespace pollution, but on a function level nothing bad happens